js/src/launchAsync.ts

```diff
@@ -63,6 +74,259 @@ let launchResponseResolve: LaunchResponseResolve;
+                };
+
+                const pauseMessageValue: any = {
+                    command: 'PlayState',
+                    parameters: 'Pause'
+                };
+
+                launchResponse = {
+                    container: iframeContainer,
+                    sessionId: response.sessionId,
+                    charactersProcessed: response.meteredContentSize,
+                    postLaunchOperations: {
+                        pause: () => {
+                            iframe.contentWindow!.postMessage(JSON.stringify({ messageType: 'InstrumentationCommand', messageValue: pauseMessageValue }), '*');
+                        },
+                        play: () => {
+                            iframe.contentWindow!.postMessage(JSON.stringify({ messageType: 'InstrumentationCommand', messageValue: playMessageValue }), '*');
+                        }
+                    }
+                };
+            } else if (response && !response.success) {
+                error = {
+                    code: response.errorCode,
+                    message: errorMessageMap[response.errorCode],
+                    sessionId: response.sessionId
+                };
+            } else {
+                error = {
+                    code: ErrorCode.ServerError,
+                    message: errorMessageMap[ErrorCode.ServerError]
+                };
+            }
+
+            isLoading = false;
+
+            if (funcType === FuncType.launchAsync) {
+                const {
+                    reject,
+                    resolve
+                } = extras;
+                if (launchResponse) {
+                    resetTimeout();
+                    resolve(launchResponse);
+                } else if (error) {
+                    exit(options.onExit);
+                    reject(error);
+                }
+            } else if (funcType === FuncType.launchWithoutContentAsync) {
+                if (launchResponse) {
+                    resetTimeout();
+                    if (launchResponseResolve) {
+                        launchResponseResolve(launchResponse);
+                    }
+                } else if (error) {
+                    exit(options.onExit);
+                    if (launchResponseReject) {
+                        launchResponseReject(error);
+                    }
+                }
+            }
+        } else if (e.data.startsWith(PostMessagePreferences)) {
+            if (options.onPreferencesChanged && typeof options.onPreferencesChanged === 'function') {
+                try {
+                    options.onPreferencesChanged(e.data.substring(PostMessagePreferences.length));
+                } catch {
+                    // no-op?
+                }
+            }
+        }
+    };
+}
+
+function _makeSrcUrl(options: Options, funcType: FuncType): string {
+    let src = '';
+    if (funcType === FuncType.launchAsync) {
+        src = 'reader?exitCallback=ImmersiveReader-Exit';
```

Comment:
I think you should build the full URL here so it is more clear and maintained in one place

Reply:
as in not have the separate additions below?

I tried to reduce code duplication there